### PR TITLE
feat: tuple deserialisation

### PR DESCRIPTION
**Pull request by lmmx**
Working on finished tuple deserialisation

- Follow up to #314

### Integration

- [x] i32 tuples
- [x] f32 tuples
- [x] i32/f32 mixed tuples
- [x] string tuple
- [x] nested tuple (tuple inside a tuple)

Also contains ignored integration/error tests for empty tuple handling, treating it as an edge case we don't handle (yet)

  - Tuple type includes unit type https://doc.rust-lang.org/reference/types/tuple.html

Snapshot test showing the error when the tuple is mis-sized could be made clearer (say "wrong size") but OK for now. Snapshot not included in this PR but can be seen if you un-ignore the 2 tests in `facet-json/tests/integration/err.rs`.

```
A: facet-json/tests/integration/snapshots/r#mod__integration__err__deserialize_tuple_wrong_size.snap
@@ -0,0 +1,11 @@
+---
+source: facet-json/tests/integration/err.rs
+expression: err
+---
+Error: Error at $
+   ╭─[ json:1:4 ]
+   │
+ 1 │ [10,20]
+   │    ┬  
+   │    ╰── Field error for shape (i32): No such static field
+───╯
```

I added another for when you read a '0-tuple' (unit type) but receive data in the `[...]`. This is giving the wrong error message for now, but this snapshot can be updated when it correctly says "hey that's not an empty container"

```
A: facet-json/tests/integration/snapshots/r#mod__integration__err__deserialize_tuple_empty.snap
@@ -0,0 +1,11 @@
+---
+source: facet-json/tests/integration/err.rs
+expression: err
+---
+Error: Error at $
+   ╭─[ json:1:1 ]
+   │
+ 1 │ [10]
+   │ ┬  
+   │ ╰── Unsupported type: got (), wanted array, list, or slice
+───╯
```

### CODE REVIEW

File: facet-core/src/types/mod.rs [hunk 1]
```diff
@@ -456,6 +456,14 @@
             _ => Err(self),
         }
     }
+    /// Returns the `TupleDef` wrapped in an `Ok` if this is a [`Def::Struct`] whose kind is
+    /// [`StructKind::Tuple`].
+    pub fn into_tuple(self) -> Result<Struct, Self> {
```
[the remark this answers is not in the document]
Comment (lmmx) [reply facet-core/src/types/mod.rs:461]: :thinking:

[the remark this answers is not in the document]
Comment (fasterthanlime) [reply facet-core/src/types/mod.rs]: yeah it's back to being called `StructDef` now.

but also.. `Tuple != Struct`? idk

[the remark this answers is not in the document]
Comment (lmmx) [reply facet-core/src/types/mod.rs]: fixed

**Comment by lmmx** [comment]
@fasterthanlime having trouble working out how to fix this last bit, handling the unit type as a 0-tuple 7038d66

**Comment by fasterthanlime** [comment]
I'm on my mobile phone, but the code looks reasonable to me. What exactly you struggling with?

**Comment by lmmx** [comment]
No worries, just am gonna be away from tomorrow so don't want to leave this half-finished !

I just need to work out how to say prematurely "that's all the elements in the tuple, close it" for the case of the unit type `()` which we are treating as a 0-tuple.

I seem to have correctly 'entered' the tuple via the `[` in the JSON but then am getting an error on reading the `]` straight after

```
louis 🌟 ~/dev/facet/facet $ just test "-p facet-json -E 'test(integration::read::tuple::test_deserialize_tuple_empty)'"
...
thread 'integration::read::tuple::test_deserialize_tuple_empty' panicked at facet-json/tests/integration/read/tuple.rs:135:12:
called `Result::unwrap()` on an `Err` value: Error: Error at $
   ╭─[ json:1:2 ]
   │
 1 │ []
   │  ┬  
   │  ╰── Wrong shape: expected list or array, but got ()
───╯
```

Screenshot from 2025-04-25 21-38-12

**Comment by lmmx** [comment]
Ah, I think I got it, now past the inside of the list - it was the pushback I hadn't updated with the unit type case :bulb: 

Screenshot from 2025-04-25 22-08-59

**Comment by lmmx** [comment]
Yeah, looks like I'm gonna have to walk this one back, can't see how to terminate it correctly :disappointed: I marked as fully initialised immediately, I don't get what I've missed. I can't pop a frame off to finish

It thinks it's reached EOF in a value when the value should have been finished (it's that "how to tell it it's finished" I am not getting)

Screenshot from 2025-04-25 23-08-37

**Comment by lmmx** [comment]
I'm sure it'll be achievable soon, but in the interest of not letting this go half baked I'm going to ignore the test

Also, I decided it's easier to interpret if you deserialize them as scalar types in a list:

```rust
Result(<(), (), _>) = from_str("[[],[]]")
```

so changed the ignored test to that

This PR introduces lots of tests for deserialising tuples and they all seem to work :-)

File: facet-json/src/deserialize.rs [hunk 2]
```diff
@@ -402,10 +402,60 @@ pub fn from_slice_wip<'input: 'facet, 'facet>(
...
                                     }
+                                    Def::Struct(s) => {
+                                        if s.kind == StructKind::Tuple {
+                                            trace!(
+                                                "Array starting for tuple ({})!",
+                                                wip.shape().blue()
+                                            );
+                                            // Special handling for unit type ()
+                                            if s.fields.is_empty() {
+                                                // Check if the array is empty by peeking at the next token
```
Comment (fasterthanlime) [inline facet-json/src/deserialize.rs:413]: This is probably a fine approach, but I think I would have also... Another approach would have been to wait until we get another value, and then when we push it we get a reflect error, and then when we get a reflect error we... I don't know, at that time we generate a nice JSON error instead?

File: facet-json/src/deserialize.rs [hunk 3]
```diff
@@ -402,10 +402,60 @@ pub fn from_slice_wip<'input: 'facet, 'facet>(
...
+                                                    // Non-empty array is not valid for unit type
+                                                    bail!(JsonErrorKind::UnsupportedType {
+                                                        got: wip.innermost_shape(),
+                                                        wanted: "empty array",
+                                                    });
+                                                }
+                                            } else {
+                                                reflect!(put_default());
+                                            }
+                                        } else {
```
Comment (fasterthanlime) [inline facet-json/src/deserialize.rs:428]: In general, I would love to get to see more early returns in there to avoid such deep indentation.

**Review by fasterthanlime — APPROVED** [review]
Looks good, I added some comments to keep in mind for further PRs.

File: facet-json/src/deserialize.rs (hunk 3 above)
Comment (lmmx) [reply facet-json/src/deserialize.rs:428]: I returned to address this but see it's been modified (refactored to remove macros: 6e1d479) which I think resolved this complaint. Specifically [here](https://github.com/lmmx/facet/blob/22af2b63dd8aa63f1b1ff750f599732503087a1a/facet-json/src/deserialize.rs#L422C1-L430C38) there is now a `?` which will early return on error :smiley_cat: :+1:
